src/core/server/integration_tests/saved_objects/migrations/zdt_2/v2_to_zdt_partial_failure.test.ts

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import Path from 'path';
+import fs from 'fs/promises';
+import { range } from 'lodash';
+import { type TestElasticsearchUtils } from '@kbn/core-test-helpers-kbn-server';
+import { SavedObjectsBulkCreateObject } from '@kbn/core-saved-objects-api-server';
+import '../jest_matchers';
+import { getKibanaMigratorTestKit, startElasticsearch } from '../kibana_migrator_test_kit';
+import { delay, parseLogFile, createType } from '../test_utils';
+import { getBaseMigratorParams, noopMigration } from '../fixtures/zdt_base.fixtures';
+
+const logFilePath = Path.join(__dirname, 'v2_to_zdt_partial_failure.test.log');
+
+describe('ZDT with v2 compat - recovering from partially migrated state', () => {
+  let esServer: TestElasticsearchUtils['es'];
+
+  beforeAll(async () => {
+    await fs.unlink(logFilePath).catch(() => {});
+    esServer = await startElasticsearch();
+  });
+
+  afterAll(async () => {
+    await esServer?.stop();
+    await delay(10);
+  });
+
+  const typeBefore = createType({
+    name: 'switching_type',
+    mappings: {
+      properties: {
+        text: { type: 'text' },
+        keyword: { type: 'keyword' },
+      },
+    },
+    migrations: {
+      '7.0.0': noopMigration,
+      '7.5.0': noopMigration,
+    },
+  });
+
+  const typeFailingBetween = createType({
+    ...typeBefore,
+    switchToModelVersionAt: '8.0.0',
+    modelVersions: {
+      1: {
+        changes: [
+          {
+            type: 'data_backfill',
+            backfillFn: (doc) => {
+              // this was the easiest way to simulate a migrate failure during doc mig.
+              throw new Error('need something to interrupt the migration');
+            },
+          },
```

Comment:
That's the most "close to reality" way to simulate the problem I've found: I just forces an error during the first zdt migration, before rerunning it and confirming that it was able to resume from the partially migrated state